README.md
# Simple Search

Provides widgets to easily list and search.

This project uses the [Infinite Scroll Pagination](https://pub.dev/packages/infinite_scroll_pagination) library to list and paginate the data.

![App](https://media.giphy.com/media/c2lNiPBETtN0iyI9xj/giphy.gif)

## Usage

Add it in your `pubspec.yaml`:

```yaml
dependencies:
  infinite_scroll_pagination: any
  simple_search: any
```

Import it.

```dart
import 'package:simple_search/simple_search.dart';
```

### SimpleSearch

All SimpleSearch tools depend on the [PagedSliverList](https://pub.dev/documentation/infinite_scroll_pagination/latest/infinite_scroll_pagination/PagedSliverList-class.html) from the [Infinite Scroll Pagination](https://pub.dev/packages/infinite_scroll_pagination) library.

```dart
    SimpleSearch<int, Person>(
        pagedSliverList: PagedSliverList(
          pagingController: _pagingController,
          builderDelegate: PagedChildBuilderDelegate(
            animateTransitions: true,
            itemBuilder: (context, item, index) => ListTile(
              title: Text(item.name),
            ),
          ),
        ),
        searchBar: SimpleSearchBar(
          searchBarElevation: 3.0,
          padding: EdgeInsets.all(24),
          leading: IconButton(
            onPressed: () {
              Navigator.of(context).pop();
            },
            icon: Icon(Icons.close),
          ),
          topLeading: true,
          title: Padding(
            padding: const EdgeInsets.all(8.0),
            child: Text(
              'Find a pokemon or people',
              style: Theme.of(context)
                  .textTheme
                  .titleMedium
                  ?.copyWith(fontWeight: FontWeight.w500),
            ),
          ),
          clearAction: Icon(Icons.delete),
          textFieldDecoration: BoxDecoration(
              color: Colors.white, borderRadius: BorderRadius.circular(8)),
          inputDecoration: InputDecoration(
            prefixIcon: Icon(Icons.search),
            border: InputBorder.none,
          ),
          onChangeSearch: (text) {
            // Does not search if it is the same text
            if (text != _searchTerm) {
              _searchTerm = text;
              _pagingController.refresh();
            }
          },
          searchTermValidator: (value) {
            if (value != null && value.length < 3) {
              return false;
            }
            return true;
          },
        ),
      ),
```

It is also possible to fix the search bar and other things

```dart
    SimpleSearch<int, Person>.persistent(
          pinnedSearchBar: true,
          floatingSearchBar: false,
          persistentSearchBar: SimpleSearchBar(
            searchBarElevation: 8.0,
            leading: IconButton(
              onPressed: () {
                Navigator.of(context).pop();
              },
              icon: Icon(Icons.arrow_back),
            ),
            clearAction: Icon(Icons.clear),
            textFieldDecoration: BoxDecoration(
                color: Colors.white, borderRadius: BorderRadius.circular(8)),
            onChangeSearch: (text) {
              _searchTerm = text;
              _pagingController.refresh();
            },
            searchTermValidator: (value) {
              if (value != null && value.length < 3) {
                return false;
              }
              return true;
            },
          ),
          pagedSliverList: PagedSliverList(
            pagingController: _pagingController,
            builderDelegate: PagedChildBuilderDelegate(
              animateTransitions: true,
              itemBuilder: (context, item, index) => ListTile(
                title: Text(item.name),
              ),
            ),
          ),
        ),
```

The search bar can have round edges for example:

```dart
        SimpleSearchBar.google( //  SimpleSearchBar.topLeading for top leading
                leading: IconButton(
                  onPressed: () {
                    Navigator.of(context).pop();
                  },
                ...
```

CHANGELOG.md
## 0.1.0

* Add top leading
* Update readme
* Update example 

## 0.0.1

* TODO: Describe initial release.


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:sample/seach_persist.dart';
import 'package:sample/search_normal.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: HomePage(title: 'Simple Search Sample'),
    );
  }
}

class HomePage extends StatelessWidget {
  final String title;
  const HomePage({Key? key, required this.title}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            ElevatedButton(
              onPressed: () {
                Navigator.of(context).push(
                  MaterialPageRoute<void>(
                    builder: (BuildContext context) => SearchDemoNormal(),
                  ),
                );
              },
              child: Text('Demo normal'),
            ),
            SizedBox.fromSize(
              size: Size.fromRadius(8),
            ),
            ElevatedButton(
              onPressed: () {
                Navigator.of(context).push(
                  MaterialPageRoute<void>(
                    builder: (BuildContext context) => SearchDemoPersist(),
                  ),
                );
              },
              child: Text('Demo persist'),
            ),
          ],
        ),
      ),
    );
  }
}


lib/src/simple_search.dart
import 'package:flutter/material.dart';
import 'package:infinite_scroll_pagination/infinite_scroll_pagination.dart';
import 'package:simple_search/src/simple_search_bar.dart';

class SimpleSearch<A, T> extends StatefulWidget {
  SimpleSearch({
    Key? key,
    required Widget searchBar,
    required this.pagedSliverList,
  })  : effectiveSearchBar = SliverToBoxAdapter(
          child: searchBar,
        ),
        super(key: key);

  SimpleSearch.persistent({
    Key? key,
    required SimpleSearchBar persistentSearchBar,
    double height = 56.0,
    double maxHeigth = 130.0,
    bool pinnedSearchBar = false,
    bool floatingSearchBar = false,
    required this.pagedSliverList,
  })  : effectiveSearchBar = SliverPersistentHeader(
          pinned: pinnedSearchBar,
          floating: floatingSearchBar,
          delegate: PersistentSearchBar(
              widget: persistentSearchBar,
              height: (persistentSearchBar.topLeading ?? false
                  ? height + 48
                  : height)),
        ),
        super(key: key);

  final Widget effectiveSearchBar;
  final PagedSliverList<A, T> pagedSliverList;

  @override
  _SimpleSearchState createState() => _SimpleSearchState();
}

class _SimpleSearchState extends State<SimpleSearch> {
  @override
  Widget build(BuildContext context) {
    return CustomScrollView(
      slivers: [
        widget.effectiveSearchBar,
        widget.pagedSliverList,
      ],
    );
  }
}

class PersistentSearchBar extends SliverPersistentHeaderDelegate {
  final Widget widget;
  final double height;
  final double? maxHeigth;

  PersistentSearchBar(
      {required this.widget, required this.height, this.maxHeigth});

  @override
  Widget build(
      BuildContext context, double shrinkOffset, bool overlapsContent) {
    return PreferredSize(
      child: widget,
      preferredSize: Size.fromHeight(height),
    );
  }

  @override
  double get maxExtent => maxHeigth ?? height;

  @override
  double get minExtent => height;

  @override
  bool shouldRebuild(SliverPersistentHeaderDelegate oldDelegate) {
    return true;
  }
}


example/lib/seach_persist.dart
import 'package:flutter/material.dart';
import 'package:infinite_scroll_pagination/infinite_scroll_pagination.dart';
import 'package:simple_search/simple_search.dart';

class SearchDemoPersist extends StatefulWidget {
  SearchDemoPersist({Key? key, this.title = 'Demo Persist'}) : super(key: key);

  final String title;

  @override
  _SearchDemoPersistState createState() => _SearchDemoPersistState();
}

class _SearchDemoPersistState extends State<SearchDemoPersist> {
  String? _searchTerm;

  final listPersons = [
    Person('Charizard'),
    Person('Pikachu'),
    Person('Afondo'),
    Person('Aristo'),
    Person('Jackin'),
    Person('Jack'),
    Person('Jacquin'),
    Person('Oliver'),
    Person('Olivia'),
    Person('Daniel'),
    Person('Dan'),
    Person('Marie'),
    Person('Maria'),
    Person('Marine'),
    Person('Marine1'),
    Person('Marine2'),
    Person('Marine3'),
    Person('Marine4'),
    Person('Marine5'),
    Person('Marine6'),
    Person('Marine7'),
    Person('Marine8'),
    Person('Marine9'),
    Person('Marine10'),
    Person('Marine11'),
    Person('Marine12'),
    Person('Marine13'),
    Person('Marine14'),
  ];

  final PagingController<int, Person> _pagingController =
      PagingController(firstPageKey: 0);

  @override
  void initState() {
    _pagingController.addPageRequestListener((pageKey) {
      _fetchPage(pageKey);
    });
    super.initState();
  }

  Future<void> _fetchPage(int pageKey) async {
    try {
      final result = await Future.value(listPersons
          .where((element) => element.name
              .toLowerCase()
              .contains(_searchTerm?.toLowerCase() ?? ''))
          .toList());

      if (_searchTerm?.isEmpty ?? true) {
        _pagingController.appendLastPage([]);
      } else {
        _pagingController.appendLastPage(result);
      }
    } catch (error) {
      _pagingController.error = error;
    }
  }

  void _incrementCounter() {
    showDialog(
      context: context,
      builder: (context) {
        return AlertDialog(
          title: Text('Search'),
          content: Container(
            width: double.maxFinite,
            child: SimpleSearch<int, Person>(
              pagedSliverList: PagedSliverList(
                pagingController: _pagingController,
                builderDelegate: PagedChildBuilderDelegate(
                  animateTransitions: true,
                  itemBuilder: (context, item, index) => ListTile(
                    title: Text(item.name),
                  ),
                ),
              ),
              searchBar: SimpleSearchBar.google(
                leading: IconButton(
                  onPressed: () {
                    Navigator.of(context).pop();
                  },
                  icon: Icon(Icons.arrow_back),
                ),
                clearAction: Icon(Icons.clear),
                textFieldDecoration: BoxDecoration(
                    color: Colors.white,
                    borderRadius: BorderRadius.circular(8)),
                onChangeSearch: (text) {
                  _searchTerm = text;
                  _pagingController.refresh();
                },
                searchTermValidator: (value) {
                  if (value != null && value.length < 3) {
                    return false;
                  }
                  return true;
                },
              ),
            ),
          ),
        );
      },
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          _incrementCounter();
        },
      ),
      body: SafeArea(
        child: SimpleSearch<int, Person>.persistent(
          maxHeigth: 160,
          height: 80,
          pinnedSearchBar: true,
          persistentSearchBar: SimpleSearchBar(
            searchBarElevation: 8.0,
            topLeading: true,
            leading: IconButton(
              onPressed: () {
                Navigator.of(context).pop();
              },
              icon: Icon(Icons.arrow_back),
            ),
            title: Padding(
              padding: const EdgeInsets.all(8.0),
              child: Text(
                'Find a pokemon or people',
                style: Theme.of(context)
                    .textTheme
                    .titleMedium
                    ?.copyWith(fontWeight: FontWeight.w500),
              ),
            ),
            clearAction: Icon(Icons.clear),
            textFieldDecoration: BoxDecoration(
                color: Colors.white, borderRadius: BorderRadius.circular(8)),
            onChangeSearch: (text) {
              _searchTerm = text;
              _pagingController.refresh();
            },
            searchTermValidator: (value) {
              if (value != null && value.length < 3) {
                return false;
              }
              return true;
            },
          ),
          pagedSliverList: PagedSliverList(
            pagingController: _pagingController,
            builderDelegate: PagedChildBuilderDelegate(
              animateTransitions: true,
              itemBuilder: (context, item, index) => ListTile(
                title: Text(item.name),
              ),
            ),
          ),
        ),
      ),
    );
  }
}

class Person {
  final String name;

  Person(this.name);
}


lib/src/simple_search_bar.dart
import 'dart:async';

import 'package:flutter/material.dart';

typedef SearchTermValidator<String> = bool Function(String? value);

class SimpleSearchBar extends StatefulWidget {
  const SimpleSearchBar({
    Key? key,
    this.leading,
    this.topLeading,
    this.padding,
    this.title,
    this.textFieldDecoration,
    this.inputDecoration,
    this.clearAction = const Icon(Icons.clear),
    required this.onChangeSearch,
    this.searchBarElevation = 0.0,
    this.searchBorderRadiusGeometry,
    this.textFieldTextStyle,
    this.debounceDuration = const Duration(milliseconds: 500),
    this.searchBarDecoration,
    this.searchTermValidator,
  }) : super(key: key);

  /// Put the leading on top of the bar
  const SimpleSearchBar.topLeading({
    Key? key,
    this.leading,
    this.topLeading = true,
    this.padding,
    this.title,
    this.textFieldDecoration,
    this.inputDecoration,
    this.clearAction = const Icon(Icons.clear),
    required this.onChangeSearch,
    this.searchBarElevation = 0.0,
    this.searchBorderRadiusGeometry,
    this.textFieldTextStyle,
    this.debounceDuration = const Duration(milliseconds: 500),
    this.searchBarDecoration,
    this.searchTermValidator,
  }) : super(key: key);

  /// Imitation of google search bar
  const SimpleSearchBar.google({
    Key? key,
    this.leading,
    this.topLeading,
    this.padding,
    this.title,
    this.textFieldDecoration,
    this.inputDecoration,
    this.clearAction = const Icon(Icons.clear),
    required this.onChangeSearch,
    this.searchBarElevation = 8.0,
    this.searchBorderRadiusGeometry =
        const BorderRadius.all(Radius.circular(48)),
    this.textFieldTextStyle,
    this.debounceDuration = const Duration(milliseconds: 500),
    this.searchBarDecoration = const BoxDecoration(
      color: Colors.white,
      borderRadius: const BorderRadius.all(Radius.circular(48)),
    ),
    this.searchTermValidator,
  }) : super(key: key);

  /// A widget to display before the search field.
  /// Typically the leading widget is an Icon or an IconButton.
  final Widget? leading;

  /// Put the leading on the top of bar
  final bool? topLeading;

  /// Title of bar
  final Widget? title;

  // Padding of the bar
  final EdgeInsetsGeometry? padding;

  /// A widget with action that clear search field.
  /// Typically the cleatAction widget is an Icon or an IconButton.
  final Widget? clearAction;

  /// The decoration of search bar.
  final Decoration? searchBarDecoration;

  /// Elevation of search bar.
  final double searchBarElevation;

  /// Border of search bar.
  final BorderRadiusGeometry? searchBorderRadiusGeometry;

  /// Define style for text in search field.
  final TextStyle? textFieldTextStyle;

  /// Decoration of text in search field.
  final Decoration? textFieldDecoration;

  /// [InputDecoration] of search field.
  final InputDecoration? inputDecoration;

  /// Debouce duration help adjust perfomance.
  final Duration? debounceDuration;

  /// Handle change of search terms.
  final void Function(String?) onChangeSearch;

  /// Determines the condition for the search to be done
  final SearchTermValidator<String>? searchTermValidator;

  @override
  _SimpleSearchBarState createState() => _SimpleSearchBarState();
}

class _SimpleSearchBarState extends State<SimpleSearchBar> {
  var _cancelIsVisible = false;
  final _searchTextFieldController = TextEditingController();
  Timer? _debounce;

  String get _value => _searchTextFieldController.value.text;
  bool get isValid => widget.searchTermValidator?.call(_value) ?? true;

  void _cancel() {
    _searchTextFieldController.clear();
  }

  void _onSearchDebounce(String text) async {
    if (_debounce?.isActive ?? false) _debounce?.cancel();
    _debounce =
        Timer(widget.debounceDuration ?? Duration(milliseconds: 500), () {
      final _seachTerm = isValid ? text : null;
      widget.onChangeSearch(_seachTerm);
      setState(() {
        _cancelIsVisible = text.isNotEmpty;
      });
    });
  }

  @override
  void initState() {
    _searchTextFieldController.addListener(() {
      _onSearchDebounce(_searchTextFieldController.text);
    });
    super.initState();
  }

  @override
  void dispose() {
    _searchTextFieldController.dispose();
    _debounce?.cancel();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final widthMax = MediaQuery.of(context).size.width;

    bool? topLeading = widget.topLeading ?? false;

    Widget? leading = widget.leading;
    if (leading != null && topLeading == false) {
      leading = ConstrainedBox(
        constraints: BoxConstraints.tightFor(width: 80),
        child: leading,
      );
    }

    EdgeInsetsGeometry? padding = widget.padding ?? EdgeInsets.all(8);

    Widget? title = widget.title;

    return Material(
      borderRadius: widget.searchBorderRadiusGeometry,
      elevation: widget.searchBarElevation,
      child: Container(
        padding: padding,
        decoration: widget.searchBarDecoration,
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            topLeading ? leading ?? SizedBox.shrink() : SizedBox.shrink(),
            title ?? SizedBox.shrink(),
            Row(
              mainAxisAlignment: MainAxisAlignment.start,
              children: <Widget>[
                !topLeading ? leading ?? SizedBox.shrink() : SizedBox.shrink(),
                Flexible(
                  child: AnimatedContainer(
                    duration: Duration(milliseconds: 200),
                    width: _cancelIsVisible ? widthMax * .9 : widthMax,
                    decoration: widget.textFieldDecoration,
                    child: Padding(
                      padding: EdgeInsets.all(8),
                      child: Theme(
                        child: TextField(
                          controller: _searchTextFieldController,
                          style: widget.textFieldTextStyle,
                          decoration: widget.inputDecoration,
                        ),
                        data: Theme.of(context).copyWith(
                          primaryColor: Colors.black,
                        ),
                      ),
                    ),
                  ),
                ),
                GestureDetector(
                  onTap: _cancel,
                  child: AnimatedOpacity(
                    opacity: _cancelIsVisible ? 1.0 : 0,
                    curve: Curves.easeIn,
                    duration:
                        Duration(milliseconds: _cancelIsVisible ? 500 : 0),
                    child: AnimatedContainer(
                      duration: Duration(milliseconds: 200),
                      width: _cancelIsVisible
                          ? MediaQuery.of(context).size.width * .1
                          : 0,
                      child: Container(
                        color: Colors.transparent,
                        child: Center(
                          child: widget.clearAction,
                        ),
                      ),
                    ),
                  ),
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}


example/lib/search_normal.dart
import 'package:flutter/material.dart';
import 'package:infinite_scroll_pagination/infinite_scroll_pagination.dart';
import 'package:simple_search/simple_search.dart';

class SearchDemoNormal extends StatefulWidget {
  SearchDemoNormal({Key? key, this.title = 'Demo normal'}) : super(key: key);

  final String title;

  @override
  _SearchDemoNormalState createState() => _SearchDemoNormalState();
}

class _SearchDemoNormalState extends State<SearchDemoNormal> {
  String? _searchTerm;

  final listPersons = [
    Person('Charizard'),
    Person('Pikachu'),
    Person('Afondo'),
    Person('Aristo'),
    Person('Jackin'),
    Person('Jack'),
    Person('Jacquin'),
    Person('Oliver'),
    Person('Olivia'),
    Person('Daniel'),
    Person('Dan'),
    Person('Marie'),
    Person('Maria'),
    Person('Marine'),
  ];

  final PagingController<int, Person> _pagingController =
      PagingController(firstPageKey: 0);

  @override
  void initState() {
    _pagingController.addPageRequestListener((pageKey) {
      _fetchPage(pageKey);
    });
    super.initState();
  }

  Future<void> _fetchPage(int pageKey) async {
    try {
      final result = await Future.value(listPersons
          .where((element) => element.name
              .toLowerCase()
              .contains(_searchTerm?.toLowerCase() ?? ''))
          .toList());

      if (_searchTerm?.isEmpty ?? true) {
        _pagingController.appendLastPage([]);
      } else {
        _pagingController.appendLastPage(result);
      }
    } catch (error) {
      _pagingController.error = error;
    }
  }

  void _incrementCounter() {
    showDialog(
      context: context,
      builder: (context) {
        return AlertDialog(
          title: Text('Search'),
          content: Container(
            width: double.maxFinite,
            child: SimpleSearch<int, Person>(
              pagedSliverList: PagedSliverList(
                pagingController: _pagingController,
                builderDelegate: PagedChildBuilderDelegate(
                  animateTransitions: true,
                  itemBuilder: (context, item, index) => ListTile(
                    title: Text(item.name),
                  ),
                ),
              ),
              searchBar: SimpleSearchBar.google(
                leading: IconButton(
                  onPressed: () {
                    Navigator.of(context).pop();
                  },
                  icon: Icon(Icons.arrow_back),
                ),
                clearAction: Icon(Icons.clear),
                textFieldDecoration: BoxDecoration(
                    color: Colors.white,
                    borderRadius: BorderRadius.circular(8)),
                onChangeSearch: (text) {
                  debugPrint('chamou');
                  _searchTerm = text;
                  _pagingController.refresh();
                },
                searchTermValidator: (value) {
                  if (value != null && value.length < 3) {
                    return false;
                  }
                  return true;
                },
              ),
            ),
          ),
        );
      },
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          _incrementCounter();
        },
      ),
      body: SimpleSearch<int, Person>(
        pagedSliverList: PagedSliverList(
          pagingController: _pagingController,
          builderDelegate: PagedChildBuilderDelegate(
            animateTransitions: true,
            itemBuilder: (context, item, index) => ListTile(
              title: Text(item.name),
            ),
          ),
        ),
        searchBar: SimpleSearchBar(
          searchBarElevation: 3.0,
          padding: EdgeInsets.all(24),
          leading: IconButton(
            onPressed: () {
              Navigator.of(context).pop();
            },
            icon: Icon(Icons.close),
          ),
          topLeading: true,
          title: Padding(
            padding: const EdgeInsets.all(8.0),
            child: Text(
              'Find a pokemon or people',
              style: Theme.of(context)
                  .textTheme
                  .titleMedium
                  ?.copyWith(fontWeight: FontWeight.w500),
            ),
          ),
          clearAction: Icon(Icons.delete),
          textFieldDecoration: BoxDecoration(
              color: Colors.white, borderRadius: BorderRadius.circular(8)),
          inputDecoration: InputDecoration(
            prefixIcon: Icon(Icons.search),
            border: InputBorder.none,
          ),
          onChangeSearch: (text) {
            // Does not search if it is the same text
            if (text != _searchTerm) {
              _searchTerm = text;
              _pagingController.refresh();
            }
          },
          searchTermValidator: (value) {
            if (value != null && value.length < 3) {
              return false;
            }
            return true;
          },
        ),
      ),
    );
  }
}

class Person {
  final String name;

  Person(this.name);
}
